content/en/blog/2022/rust-instrumentation/index.md

```diff
@@ -0,0 +1,401 @@
+
+Spans - In contrast to a log line, which represents a precise moment in time, a span denotes a period of time with a beginning and an end.
+
+Events - An Event could be compared to the log records generated by unstructured logging code, but unlike a typical log line, Events may occur within spans.
+
+To record spans and events we use macros, here are some examples:
+
+```rust
+{
+    // constructs a span with TRACE level.
+    let span = span!(Level::TRACE, "my span");
+    let _enter_guard = span.enter();
+
+    // recording an INFO level event
+    event!(Level::INFO, "inside the block");
+
+    // the _enter_guard will be dropped here and the span will exit.
+}
+{
+    let some_veriable = 100_000;
+    // the entered consumes the span. which means it now owns it instead of just borrowing it.
+    let _enter_guard = info_span!("INFO level span").entered();
+
+    // records an ERROR level event with some attributes
+    error!(it = "accepts multiple arguments", and_also = ("multiple types"), how_many="?", could_be_more_than=some_veriable);
+}
+```
+
+Subscriber - For spans and events to be recorded and aggregated we need to initialize a Subscriber. In most cases, we will use an existing Subscriber implementation.
+
+The crate “[tracing_subscriber](https://docs.rs/tracing-subscriber/0.3.15/tracing_subscriber/)” provides us with general use case subscriber that can be extended and modified according to specific needs.
+
+We initialize the subscriber using Registry struct:
+
+```shell
+let subscriber = tracing_subscriber::Registry::default()
+```
+
+This struct implements a subscriber and exposes us to another important feature of this crate - The Layer. Using layers, we can configure our subscriber to apply specific behaviors when interacting with spans and events.
+
+For example, if we want to filter some of our data, export it, format it or edit it, we can create or use an existing layer and compose it with the registry subscriber.
+
+```shell
+let subscriber = tracing_subscriber::Registry::default().with(SomeLayer)
+```
+
+We will see more examples on how to use these crates to instrument a Rust program in the practical section.
+
+### OpenTelemetry in Rust
+
+The crate that provides support for OpenTelemetry is called - how convenient - “[OpenTelemetry](https://docs.rs/opentelemetry/latest/opentelemetry/)”.
+
+“OpenTelemetry provides a single set of APIs, libraries, agents, and collector services to capture distributed traces and metrics from your application.”
+
+Let’s go over some of its key APIs:
+
+- Tracer. Inside the crate, there is a module called traces. It introduces us to a trait called Tracer. This is what tracks and connects our spans to create traces. To use Tracer, we create a subscriber layer. This is how the subscriber knows where to send the spans and how to generate traces.
```

Comment:
@svrnm  For modules other code references, should the form `traces` be used?

Reply:
yes, makes it easier to see

Reply:
> there is a module called traces

This use of the word "traces" is different from this use of the word:

> how to generate traces.

And so if you make it visually distinct either via a link, or markdown code formatting, etc. it will be more clear.

Reply:
@NavehMevorach 
The way to fix this would be that when you're referring to the model encapsulate it
```suggestion
- Tracer. Inside the crate, there is a module called `traces`. It introduces us to a trait called Tracer. This is what tracks and connects our spans to create traces. To use Tracer, we create a subscriber layer. This is how the subscriber knows where to send the spans and how to generate traces.
```